staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -307,19 +473,166 @@ func (q *delayingType) waitingLoop() {
+		q.headReadyTime = time.Time{}
+	}
+
+	// create new timer
+	if head != nil {
+		now := q.clock.Now()
+		q.nextReadyAtTimer = q.clock.NewTimer(head.readyAt.Sub(now))
+		q.nextReadyAtTimerCh = q.nextReadyAtTimer.C()
+		q.headReadyTime = head.readyAt
+	}
+}
+
+// handleReportingAndSync loops through the inspection requests and syncs and loops through providing the answers
+// and closing the return channel.
+func (q *delayingType) handleReportingAndSync(w *waitForPriorityQueue, knownEntries map[t]*waitFor) {
+	if len(q.afterFill) == 0 {
+		return
+	}
+
+	for _, req := range q.afterFill {
+		r := waitingLoopResponse{}
+		switch req.action {
+		case waitForActionIsWaiting:
+			_, exists := knownEntries[req.data]
+			r.exists = &exists
+			nr := req.readyAt.Sub(q.clock.Now())
+			r.nextReady = &nr
+		case waitForActionLenWaiting:
+			l := w.Len()
+			r.length = &l
+		case waitForActionNextReady:
+			if w.Len() > 0 {
+				r.item = w.Peek().(*waitFor).data
+				nr := q.headReadyTime.Sub(q.clock.Now())
+				r.nextReady = &nr
+			}
+		case waitForActionSyncFill:
+		}
+		req.returnCh <- r
+		close(req.returnCh)
+	}
+	q.afterFill = nil
+}
+
+// processArrivingWaitForEntry receives messages received from clients a takes the appropriate queuing based
+// upon the entry.action requested.
+// When performing the waitForActionQueue action it behaves as follows:
+//   - An entry is dropped if it is already actively queued (much like Add() on the Interface)
+//   - Delayed items are inserted into the 'waiting' queue according to their Waiting option (TakeShorter by default)
+//   - Immediate items will be dropped when the TakeLonger Waiting option is set and they are already present in 'waiting'.
+//   - Immediate items will pre-empt/remove items in the 'waiting' queue unless the PermitActiveAndWaiting option is set.
+func (q *delayingType) processArrivingWaitForEntry(w *waitForPriorityQueue, knownEntries map[t]*waitFor, entry *waitFor) {
+	// handle the different message actions
+	switch entry.action {
+	case waitForActionIsWaiting, waitForActionLenWaiting, waitForActionNextReady, waitForActionSyncFill:
+		// all of these actions are processed in the call handleReportingAndSync()
+		if entry.returnCh == nil {
+			return
+		}
+		q.afterFill = append(q.afterFill, entry)
+		return
+	case waitForActionDoneWaiting:
+		// remove entry from the 'waiting' queue
+		existing, exists := knownEntries[entry.data]
+		if exists {
+			// remove the item from the 'waiting' queue
+			heap.Remove(w, existing.index)
+			delete(knownEntries, entry.data)
+		}
+		return
+	case waitForActionQueue:
+		// add entry to either 'active' or 'waiting' queues (dependent on delay and settings)
+		if entry.returnCh != nil {
+			defer close(entry.returnCh)
+		}
+
+		// Perform de-duplication between the 'waiting' queue against items in the 'active' queue unless
+		// disabled by PermitActiveAndWaiting.
+		if !entry.options.PermitActiveAndWaiting && q.Interface.IsQueued(entry.data) {
```

Comment:
how does this play with the `TakeLonger` or `TakeIncoming` feature? It might be a bit surprising that these items do not get in if you forget to set  `PermitActiveAndWaiting`. Any idea how to make this simpler/more straightforward?

Also, semantically it might  be a better fit to set `PermitActiveAndWaiting` on the whole delaying queue instead of per item.

For example, you might have one item in the active queue and one item in the waiting queue. And if you send the same item with `PermitActiveAndWaiting=false`, then it will not deduplicate the already existing items in the queue, so it depends on the order of operations. So the queue may not behave consistently. This would be fixed if we could only set this when initializing the delaying queue.

Reply:
TakeLonger and TakeIncome are options for insertion into the "waiting" queue but this de-duplication is between the "active" queue and "waiting" queue.  Are you thinking it would be better to somehow merge these together?  Would you like TakeShorter skip an item if it is in the "active" queue?  And would you like TakeLonger to remove an item from the "active" queue - that would be quite difficult because there is no ForgetItem(item) method on the active queue.  My gut feeling is to keep these separate.

> Also, semantically it might be a better fit to set PermitActiveAndWaiting on the whole delaying queue instead of per item

I remember having this conversation in reverse with @lavalamp at the beginning of this PR.  In my original implementation the characteristics of the delaying queue were set when you constructed the queue.  He preferred that the client could choose their behaviour through their choice of method call.

I agree that a client could get confused and use incorrectly, but in general if they use the legacy method AddAfter() then they can expect no de-duping between "active" and "waiting" and if they use AddWithOptions() they should expect de-duping by default but can be disabled with the PermitActiveWaiting option.

Reply:
> TakeLonger and TakeIncome are options for insertion into the "waiting" queue but this de-duplication is between the "active" queue and "waiting" queue. Are you thinking it would be better to somehow merge these together? Would you like TakeShorter skip an item if it is in the "active" queue? And would you like TakeLonger to remove an item from the "active" queue - that would be quite difficult because there is no ForgetItem(item) method on the active queue. My gut feeling is to keep these separate.

I agree let's keep them separate. It is better to keep the implementation of Queue simple.

> I remember having this conversation in reverse with @lavalamp at the beginning of this PR. In my original implementation the characteristics of the delaying queue were set when you constructed the queue. He preferred that the client could choose their behaviour through their choice of method call.

> I agree that a client could get confused and use incorrectly, but in general if they use the legacy method AddAfter() then they can expect no de-duping between "active" and "waiting" and if they use AddWithOptions() they should expect de-duping by default but can be disabled with the PermitActiveWaiting option.

Ok, since we have the original behaviour now in `AddAfter,` the user is full in control what is happening in the active and the waiting queue. Since they are both the owner and a consumer of their queue,  they can decide how to use it and how each call will affect the state of both queues.

Reply:
Have made sure that AddAfter sets the PermitActiveWaiting to true.
Leaving PermitActiveAndWaiting as is